client/chrome-extension/src/test/IntegrationTest.js

```diff
@@ -78,19 +90,82 @@ describe('integration between server and client', () => {
     });
   });
 
-  describe('the client', function () {
+  describe('one client', function () {
     this.timeout(20000);
-    let serverProcess;
-    beforeEach(async () => {
-      serverProcess = await startServer();
+    withRunningServer();
+    let client;
+    it('can make a connection', async () => {
+      client = new Client('ws://localhost:4000/socket', 'user');
+      await client.connectSync();
     });
-    afterEach(async () => {
-      await killServer(serverProcess);
+    it('can disconnect', async () => {
+      await client.disconnectSync();
     });
-    it('can make a connection', async () => {
-      const client = new Client('ws://localhost:4000/socket', 'user');
-      await new Promise((resolve, reject) => {
-        client.connect(resolve, reject);
+    it('gracefully handles disconnecting while unconnected', async () => {
+      await client.disconnectSync();
+      await client.disconnectSync();
+    });
+    it('gracefully handles connecting while already connected', async () => {
+      await client.connectSync();
+      await client.connectSync();
+    });
+  });
+
+  describe('two clients', function () {
+    this.timeout(20000);
+    let client1, client2;
+    withRunningServer(() => {
+      before(async () => {
+        client1 = new Client('ws://localhost:4000/socket', 'user1');
+        client2 = new Client('ws://localhost:4000/socket', 'user2');
+        await client1.connectSync();
+        client1.resetCallbacks();
+      });
+      after(async () => {
+        await client1.disconnectSync();
+        await client2.disconnectSync();
+      });
+    });
+    it('can detect when each other joins', (done) => {
+      expect(client1.prevState.group_size).to.equal(1);
+      client1.onUserJoin((newGroupSize) => {
+        // TODO this becomes an "uncaught exception" and does not actually fail the test
+        //      but done() is never called, so it hangs until the timeout
```

Comment:
Does this mean the `'state_change'` event maybe isn't being sent?

Reply:
no, this is actually referring to something specific to the test infrastructure here. When an `expect()` call fails, it throws an exception and the testrunner catches it and reports it appropriately. For some reason, when `expect()` fails within this scope (the lambda passed to `onUserJoin`), it just logs an uncaught exception and doesn't fail the test. More of a nuisance than anything, since there's a 20s timeout and that would lead to a failure in automation.

Reply:
Ah, ok... annoying